libs/template/experimental/virtual-scrolling/src/lib/model.ts

```diff
@@ -0,0 +1,179 @@
+  defaultStrategyName: string;
+  errorHandler?: ErrorHandler;
+}
+
+export const enum ListTemplateChangeType {
+  insert,
+  remove,
+  move,
+  update,
+  context,
+}
+// [value, index, oldIndex?]
+export type ListTemplateChangePayload<T> = [T, number?, number?];
+export type ListTemplateChange<T = any> = [
+  ListTemplateChangeType,
+  ListTemplateChangePayload<T>
+];
+export type ListTemplateChanges<T = any> = [
+  ListTemplateChange<T>[], // changes to apply
+  boolean // notify parent
+];
+
+export interface ListRange {
+  start: number;
+  end: number;
+}
+
+/**
+ * @Directive RxVirtualScrollStrategy
+ *
+ * @description
+ * Abstract implementation for the actual implementations of the ScrollStrategies
+ * being consumed by `*rxVirtualFor` and `rx-virtual-scroll-viewport`.
+ *
+ * This is one of the core parts for the virtual scrolling implementation. It has
+ * to determine the `ListRange` being rendered to the DOM as well as managing
+ * the layouting task for the `*rxVirtualFor` directive.
+ *
+ * @docsCategory RxVirtualFor
+ * @docsPage RxVirtualFor
+ * @publicApi
+ */
+@Directive()
+export abstract class RxVirtualScrollStrategy<
+  T,
+  U extends NgIterable<T> = NgIterable<T>
+> {
+  /** Emits when the index of the first element visible in the viewport changes. */
+  /** @internal */
+  abstract scrolledIndex$: Observable<number>;
+  /** @internal */
+  abstract renderedRange$: Observable<ListRange>;
+  /** @internal */
+  abstract contentSize$: Observable<number>;
+
+  /**
+   * @description
+   *
+   * Emits whenever an update to a single view was rendered
+   */
+  readonly viewRenderCallback = new Subject<{
+    view: EmbeddedViewRef<RxVirtualForViewContext<T, U>>;
+    item: T;
+    index: number;
+  }>();
+
+  /** @internal */
+  private nodeIndex?: number;
+
+  /** @internal */
+  protected getElement(
+    view: EmbeddedViewRef<RxVirtualForViewContext<T, U>>
+  ): HTMLElement {
+    if (this.nodeIndex !== undefined) {
+      return view.rootNodes[this.nodeIndex];
+    }
+    const rootNode = view.rootNodes[0];
+    this.nodeIndex = rootNode instanceof HTMLElement ? 0 : 1;
+    return view.rootNodes[this.nodeIndex] as HTMLElement;
+  }
+
+  /**
+   * Attaches this scroll strategy to a viewport.
+   * @param viewport The viewport to attach this strategy to.
+   * @param viewRepeater The viewRepeater attached to the viewport.
+   */
+  abstract attach(
+    viewport: RxVirtualScrollViewport,
+    viewRepeater: RxVirtualViewRepeater<any>
+  ): void;
+
+  /** Detaches this scroll strategy from the currently attached viewport. */
+  abstract detach(): void;
+
+  /**
+   * Scroll to the offset for the given index.
+   * @param index The index of the element to scroll to.
+   * @param behavior The ScrollBehavior to use when scrolling.
+   */
+  abstract scrollToIndex(index: number, behavior?: ScrollBehavior): void;
+}
+
+/** @internal */
+@Directive()
+export abstract class RxVirtualScrollViewport {
```

Comment:
What about using an interface here instead?

Reply:
not working, because it is used as DI token: 
![image](https://user-images.githubusercontent.com/4904455/232228883-a79a53c2-499c-4d37-8c16-081fab03ea40.png)